spire-flex/templates/agent-daemonset.yaml

```diff
@@ -14,6 +14,13 @@ spec:
     spec:
       containers:
         - name: {{ $.Release.Name }}-agent
+          image: {{ join "" (list 
+                   (coalesce ((.Values.agent).image).registry (.Values.image).registry "ghcr.io")
+                   (ternary "" ":" (empty (coalesce ((.Values.agent).image).registryPort (.Values.image).registryPort)))
+                   (coalesce ((.Values.agent).image).registryPort (.Values.image).registryPort)
+                   "/" (coalesce ((.Values.agent).image).name "spiffe/spire-agent") ":"
+                   (coalesce ((.Values.agent).image).tag (.Values.image).tag .Chart.AppVersion)
+                 ) | quote }}
```

Comment:
I understand what your trying to do. I'm still skeptical of avoiding values and doing it this way, but will still go along with it, somewhat reluctantly. There is some promise to the approach, but will be very foreign to most deployers, IMO.

That all being said, one thing I think would help here is that this logic I believe will be used repeatedly. It might be much more readable in the template at this location, along with being more dry, if it was placed in a macro? It could look something like:
```
{{ include "spire-flex.image" .Values "agent" "ghcr.io" "spiffe/spire-agent" .Chart.AppVersion }}
```

Reply:
I think putting this in a macro is a good idea.  I just didn't get around to it.

I also thought that the "coalesce" section might be a good macro, as it is repeated so many times, but it would then need to take a string and build up the nested parenthesis for me to feel the macro had enough value to exist.